test/micro/org/openjdk/bench/vm/compiler/pea/HashMapBench.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package org.openjdk.bench.vm.compiler.pea;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.concurrent.TimeUnit;
+import java.util.HashMap;
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 10, time = 1, timeUnit = TimeUnit.SECONDS)
+@State(Scope.Benchmark)
+@Fork(value = 3)
+public class HashMapBench {
+    @Param("100000")
+    private int size;
+
+    @Param("2")
+    private int fillInterval;
+
+    HashMap<Integer, Object> map;
+
+    @Setup
+    public void setUp() {
+        map = new HashMap<>();
+        for (int i = 0; i < size; i += fillInterval) {
+            map.put(i, new Object());
+        }
+    }
+
+    @Benchmark
+    public void replace() {
+        for (int i = 0; i < size; ++i) {
+            map.replace(i, new Object());
```

Comment:
if fillInterval=2, we should have 50% odd to get away of allocation here, right? I think allocation rate should drop 50%.  why did you say 25%?

Reply:
I'm not sure. At smaller sizes such as `-p size=100`, I see exactly 50% allocation reduction. However, with larger sizes, it looks more like 25%. I think its possible there is some behind the scenes allocations. For example, HashMap may have to allocate more auxiliary data the more real data it is holding.

Reply:
we only need 'new object()' if the corresponding internal node is found. I feel that no collateral allocation is in your macro-benchmark.  Can we consult Aleksey on this?

Reply:
micro-benchmark.

Reply:
Let me look into the HashMap implementation and profiling. It is possible that HashMap allocates more objects when there are more keys/object collisions. For example, in chaining, more object collisions results in more linked list nodes and heap allocations.